#!/bin/bash -l 
#PBS -l walltime=72:00:00,mem=62gb,nodes=2:ppn=20
#PBS -m abe 
#PBS -M llofgren@umn.edu

cd /scratch.global/llofgren/S_subaureus

module load bowtie2 

#index the ‘references’ uses the format : bowtie2-build -f [reference file] [output name]
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_ITS_IR.fasta S_subaureus_ITS_ref
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_LSU_IR.fasta S_subaureus_LSU_ref
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_GAPDH_IR.fasta S_subaureus_GAPDH_ref
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_RPB2_IR.fasta S_subaureus_RPB2_ref
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_GH63_IR.fasta S_subaureus_GH63_ref
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_RPB1_IR.fasta S_subaureus_RPB1_ref
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_ELF1_IR.fasta S_subaureus_ELF1_ref
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_G6PDH_IR.fasta S_subaureus_G6PDH_ref
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_MCM7_IR.fasta S_subaureus_MCM7_ref
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_MLS_IR.fasta S_subaureus_MLS_ref
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_LYS2_IR.fasta S_subaureus_LYS2_ref
bowtie2-build -f /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_TOP2_IR.fasta S_subaureus_TOP2_ref


#add argument -p 20 to increase the number of processors used for this command
#aligment uses the format: bowtie2 [options]* -x <bt2-idx> {-1 <m1> -2 <m2> | -U <r>} [-S <sam>]
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_ITS_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_ITS.sam
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_LSU_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_LSU.sam
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_GAPDH_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_GAPDH.sam
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_RPB2_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_RPB2.sam
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_GH63_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_GH63.sam
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_RPB1_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_RPB1.sam
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_ELF1_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_ELF1.sam
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_G6PDH_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_G6PDH.sam
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_MCM7_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_MCM7.sam
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_MLS_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_MLS.sam
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_LYS2_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_LYS2.sam
bowtie2 --very-sensitive-local -p 20 -x S_subaureus_TOP2_ref -1 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_1.fastq -2 /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_sub_combined_Pbio_to_ill_reads_2.fastq -S S_subaureus_TOP2.sam


module unload bowtie2

module load samtools/1.3

#convert SAM to BAM
samtools view -b S_subaureus_ITS.sam > S_subaureus_ITS.bam
samtools view -b S_subaureus_LSU.sam > S_subaureus_LSU.bam
samtools view -b S_subaureus_GAPDH.sam > S_subaureus_GAPDH.bam
samtools view -b S_subaureus_RPB2.sam > S_subaureus_RPB2.bam
samtools view -b S_subaureus_GH63.sam > S_subaureus_GH63.bam
samtools view -b S_subaureus_RPB1.sam > S_subaureus_RPB1.bam
samtools view -b S_subaureus_ELF1.sam > S_subaureus_ELF1.bam
samtools view -b S_subaureus_G6PDH.sam > S_subaureus_G6PDH.bam
samtools view -b S_subaureus_MCM7.sam > S_subaureus_MCM7.bam
samtools view -b S_subaureus_MLS.sam > S_subaureus_MLS.bam
samtools view -b S_subaureus_LYS2.sam > S_subaureus_LYS2.bam
samtools view -b S_subaureus_TOP2.sam > S_subaureus_TOP2.bam


#sort BAM files
samtools sort -T S_subaureus_tempITS  -o S_subaureus_ITS_sorted.bam -O bam S_subaureus_ITS.bam
samtools sort -T S_subaureus_tempLSU  -o S_subaureus_LSU_sorted.bam -O bam S_subaureus_LSU.bam
samtools sort -T S_subaureus_tempGAPDH  -o S_subaureus_GAPDH_sorted.bam -O bam S_subaureus_GAPDH.bam
samtools sort -T S_subaureus_tempRPB2  -o S_subaureus_RPB2_sorted.bam -O bam S_subaureus_RPB2.bam
samtools sort -T S_subaureus_tempGH63  -o S_subaureus_GH63_sorted.bam -O bam S_subaureus_GH63.bam
samtools sort -T S_subaureus_tempRPB1  -o S_subaureus_RPB1_sorted.bam -O bam S_subaureus_RPB1.bam
samtools sort -T S_subaureus_tempELF1  -o S_subaureus_ELF1_sorted.bam -O bam S_subaureus_ELF1.bam
samtools sort -T S_subaureus_tempG6PDH  -o S_subaureus_G6PDH_sorted.bam -O bam S_subaureus_G6PDH.bam
samtools sort -T S_subaureus_tempMCM7  -o S_subaureus_MCM7_sorted.bam -O bam S_subaureus_MCM7.bam
samtools sort -T S_subaureus_tempMLS  -o S_subaureus_MLS_sorted.bam -O bam S_subaureus_MLS.bam
samtools sort -T S_subaureus_tempLYS2  -o S_subaureus_LYS2_sorted.bam -O bam S_subaureus_LYS2.bam
samtools sort -T S_subaureus_tempTOP2  -o S_subaureus_TOP2_sorted.bam -O bam S_subaureus_TOP2.bam


##to get depth
#calculate depth on whole regions - note -a tag and -q0, needed to process p-bio converted reads
samtools depth -d 1000000 -a -q 0 S_subaureus_ITS_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_ITS_Q0_a.txt
samtools depth -d 1000000 -a -q 0 S_subaureus_LSU_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_LSU_Q0_a.txt
samtools depth -d 1000000 -a -q 0 S_subaureus_GAPDH_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_GAPDH_Q0_a.txt
samtools depth -d 1000000 -a -q 0 S_subaureus_RPB2_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_RPB2_Q0_a.txt
samtools depth -d 1000000 -a -q 0 S_subaureus_GH63_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_GH63_Q0_a.txt
samtools depth -d 1000000 -a -q 0 S_subaureus_RPB1_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_RPB1_Q0_a.txt
samtools depth -d 1000000 -a -q 0 S_subaureus_ELF1_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_ELF1_Q0_a.txt
samtools depth -d 1000000 -a -q 0 S_subaureus_G6PDH_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_G6PDH_Q0_a.txt
samtools depth -d 1000000 -a -q 0 S_subaureus_MCM7_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_MCM7_Q0_a.txt
samtools depth -d 1000000 -a -q 0 S_subaureus_MLS_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_MLS_Q0_a.txt
samtools depth -d 1000000 -a -q 0 S_subaureus_LYS2_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_LYS2_Q0_a.txt
samtools depth -d 1000000 -a -q 0 S_subaureus_TOP2_sorted.bam > /home/kennedyp/llofgren/genomes/Suillus_genomes/S_subaureus/S_subaureus_depth_TOP2_Q0_a.txt


#to remove the temp file
rm -rf /scratch.global/llofgren/S_subaureus
